doc/esm-vs-cjs.md

```diff
@@ -0,0 +1,132 @@
+The only currently supported loader hook is `@opentelemetry/instrumentation/hook.mjs`.
+
+Experimental loader is intended to be deprecated, and will be replaced with something like `--import=@opentelemetry/instrumentation/hook.mjs`
+
+Regarding `--experimental-loader` per [Node.js docs](https://nodejs.org/api/cli.html#--experimental-loadermodule):
+
+> This flag is discouraged and may be removed in a future version of Node.js. Please use `--import` with `register()` instead.
+
+<!--
+TODO
+import * as module from 'module'
+
+module.register('@opentelemetry/instrumentation/hook.mjs', import.meta.url)
+-->
+
+## Examples
+
+### Example Written in JavaScript as CJS
+
+```javascript
+/*telemetry.cjs*/
+const { NodeSDK } = require('@opentelemetry/sdk-node');
+const { ConsoleSpanExporter } = require('@opentelemetry/sdk-trace-node');
+const {
+  getNodeAutoInstrumentations,
+} = require('@opentelemetry/auto-instrumentations-node');
+
+const sdk = new NodeSDK({
+  traceExporter: new ConsoleSpanExporter(),
+  instrumentations: [getNodeAutoInstrumentations()],
+});
+
+sdk.start();
+```
+
+Startup command:
+
+```sh
+node --require ./telemetry.cjs app.js
+```
+
+### Example Written in JavaScript as ESM or TypeScript
+
+```typescript
+/*telemetry.ts | telemetry.mjs*/
+import { NodeSDK } from '@opentelemetry/sdk-node';
+import { ConsoleSpanExporter } from '@opentelemetry/sdk-trace-node';
+import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
+
+const sdk = new NodeSDK({
+  traceExporter: new ConsoleSpanExporter(),
+  instrumentations: [getNodeAutoInstrumentations()],
+});
+
+sdk.start();
+```
+
+Startup command for compiled CJS:
+
+```sh
+node --require ./telemetry.js app.js
+```
+
+Startup command for compiled ESM:
+
+```sh
+node --require ./telemetry.js --experimental-loader=@opentelemetry/instrumentation/hook.mjs app.js
+```
+
+To use `ts-node` to run the uncompiled TypeScript code, the module must be CJS.
+To use `tsx` to run the uncompiled TypeScript code as ESM, the `--import` flag must be used.
+
+### ESM Options for Different Versions of Node.js
+
+The entire startup command should include the following `NODE_OPTIONS`:
+
+| Node.js Version   | NODE_OPTIONS                                                                              |
+| ----------------- | ----------------------------------------------------------------------------------------- |
+| >=16.0.0          | `--require ./telemetry.cjs --experimental-loader=@opentelemetry/instrumentation/hook.mjs` |
+| >=18.1.0 <18.19.0 | `--require ./telemetry.cjs --experimental-loader=@opentelemetry/instrumentation/hook.mjs` |
+| >=18.19.0         | `--import ./telemetry.mjs --experimental-loader=@opentelemetry/instrumentation/hook.mjs`  |
+| 20.x              | `--import ./telemetry.mjs --experimental-loader=@opentelemetry/instrumentation/hook.mjs`  |
+| 22.x              | `--import ./telemetry.mjs --experimental-loader=@opentelemetry/instrumentation/hook.mjs`  |
```

Comment:
```suggestion
| 16.x              | `--require ./telemetry.cjs --experimental-loader=@opentelemetry/instrumentation/hook.mjs` |
| >=18.1.0 <18.19.0 | `--require ./telemetry.cjs --experimental-loader=@opentelemetry/instrumentation/hook.mjs` |
| ^18.19.0          | `--import ./telemetry.mjs --experimental-loader=@opentelemetry/instrumentation/hook.mjs`  |
| ^20.6.0           | `--import ./telemetry.mjs --experimental-loader=@opentelemetry/instrumentation/hook.mjs`  |
| 22.x              | `--import ./telemetry.mjs --experimental-loader=@opentelemetry/instrumentation/hook.mjs`  |
```

I suppose I left out the `>=20.0.0 <20.6.0` case.

Reply:
Is the `^20.6.0` case a known cutoff or just not something you have tried? I just tested v20.0.0 and 20.3.0 and 20.9.0 and they all seem to work with this start command.

Reply:
Ah, my bad. `^20.6.0` is needed for `module.register(...)`. https://nodejs.org/en/blog/release/v20.6.0#new-nodemodule-api-register-for-module-customization-hooks-new-initialize-hook

But we're not talking about module.register() here yet.
I think we *do* want to limit our supported Node.js versions *for ESM support* to `^20.6.0` if we do add a statement about supported node.js versions -- on the expectation that we'll be moving official recommendation to `module.register(...)` usage.